dot/network/notifications.go

```diff
@@ -288,21 +288,9 @@ func (s *Service) sendData(peer peer.ID, hs Handshake, info *notificationsProtoc
 		return
 	}
 
-	if s.host.messageCache != nil {
-		added, err := s.host.messageCache.put(peer, msg)
-		if err != nil {
-			logger.Errorf("failed to add message to cache for peer %s: %s", peer, err)
-			return
-		}
-
-		// TODO: ensure grandpa stores *all* previously received votes and discards them
-		// only when they are for already finalised rounds; currently this causes issues
-		// because a vote might be received slightly too early, causing a round mismatch err,
-		// causing grandpa to discard the vote. (#1855)
-		_, isConsensusMsg := msg.(*ConsensusMessage)
-		if !added && !isConsensusMsg {
-			return
-		}
+	if (s.host.messageCache != nil) && (s.host.messageCache.exists(peer, msg)) {
```

Comment:
I don't think you need parentheses here

```suggestion
	if s.host.messageCache != nil && s.host.messageCache.exists(peer, msg) {
```

Also a little trick you could have is have the `exists` method check if its receiver (`messageCache`) is nil and return false in this case. That way even if messageCache is nil, `messageCache.exists(...)` would return false and not panic. This is used in proto generated Go code for example.

Reply:
I would rather have the nil check here than in exists method. I am assuming that messageCache is not supposed to be nil.